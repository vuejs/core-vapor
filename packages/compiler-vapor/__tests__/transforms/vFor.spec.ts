import { makeCompile } from './_utils'
import {
  type ForIRNode,
  IRNodeTypes,
  transformElement,
  transformInterpolation,
  transformVBind,
  transformVFor,
  transformVOn,
} from '../../src'
import { NodeTypes } from '@vue/compiler-dom'

const compileWithVFor = makeCompile({
  nodeTransforms: [transformInterpolation, transformVFor, transformElement],
  directiveTransforms: {
    bind: transformVBind,
    on: transformVOn,
  },
})

describe('compiler: v-for', () => {
  test('basic v-for', () => {
    const { code, ir, vaporHelpers, helpers } = compileWithVFor(
      `<div v-for="item of items" @click="remove(item)">{{ item }}</div>`,
    )

    expect(code).matchSnapshot()
    expect(vaporHelpers).contains('createFor')
    expect(helpers.size).toBe(0)
    expect(ir.template).lengthOf(2)
    expect(ir.template).toMatchObject([
      {
        template: '<div></div>',
        type: IRNodeTypes.TEMPLATE_FACTORY,
      },
      {
        type: IRNodeTypes.FRAGMENT_FACTORY,
      },
    ])
    expect(ir.operation).toMatchObject([
      {
        type: IRNodeTypes.FOR,
        id: 1,
        source: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: 'items',
        },
        value: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: 'item',
        },
        key: undefined,
        index: undefined,
        render: {
          type: IRNodeTypes.BLOCK_FUNCTION,
          templateIndex: 0,
        },
      },
      {
        type: IRNodeTypes.APPEND_NODE,
        elements: [1],
        parent: 0,
      },
    ])
    expect(ir.dynamic).toMatchObject({
      id: 0,
      children: { 0: { id: 1 } },
    })
    expect(ir.effect).toEqual([])
    expect((ir.operation[0] as ForIRNode).render.effect).lengthOf(1)
  })

  test('multi effect', () => {
    const { code } = compileWithVFor(
      `<div v-for="(item, index) of items" :item="item" :index="index" />`,
    )
    expect(code).matchSnapshot()
  })

  test('w/o value', () => {
    const { code } = compileWithVFor(`<div v-for=" of items">item</div>`)
    expect(code).matchSnapshot()
  })

  test.todo('object de-structured value', () => {
    const { code } = compileWithVFor(
      '<span v-for="({ id, value }) in items">{{ id }}{{ value }}</span>',
    )
    expect(code).matchSnapshot()
  })
})
