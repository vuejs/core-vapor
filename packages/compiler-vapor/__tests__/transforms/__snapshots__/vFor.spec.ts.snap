// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: v-for > basic v-for 1`] = `
"import { template as _template, fragment as _fragment, children as _children, on as _on, setText as _setText, renderEffect as _renderEffect, createFor as _createFor, append as _append } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template("<div></div>")
  const t1 = _fragment()
  const n0 = t1()
  const n1 = _createFor(() => (_ctx.items), (_block) => {
    const n2 = t0()
    const { 0: [n3],} = _children(n2)
    _on(n3, "click", $event => (_ctx.remove(_block.s[0])))
    const _updateEffect = () => {
      const [item] = _block.s
      _setText(n3, item)
    }
    _renderEffect(_updateEffect)
    return [n2, _updateEffect]
  })
  _append(n0, n1)
  return n0
}"
`;

exports[`compiler: v-for > basic v-for 2`] = `
"import { template as _template, fragment as _fragment, createFor as _createFor, append as _append } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template("<div>item</div>")
  const t1 = _fragment()
  const n0 = t1()
  const n1 = _createFor(() => (_ctx.items), (_block) => {
    const n2 = t0()
    return [n2, () => {}]
  })
  _append(n0, n1)
  return n0
}"
`;
